### PR TITLE
Market Service - updated Market configuration properties.

### PR DESCRIPTION
**Pull request by kartaggen**
- Added file class for getting file (or creating if needed) and added to market pom.
- Updated market pom to use spring build plugin.
- Updated MarketConfiguration:
1. Implemented a loading method from file. 
2. Updated file placement due to not working when compiled to jar.
3. Added validation to product properties modification.
4. Updated GoodSpecification to not include origin.
5. Updated MarketConfigurationTest

### CODE REVIEW

File: common/src/main/java/com/market/banica/common/util/ApplicationDirectory.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package com.market.banica.common.util;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class ApplicationDirectory {
```
Comment (trayan-muchev) [inline common/src/main/java/com/market/banica/common/util/ApplicationDirectory.java:9]: You can name it ApplicationDirectoryUtil

File: common/src/main/java/com/market/banica/common/util/ApplicationDirectory.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    private static final Path config;
+
+    private ApplicationDirectory() {
+    }
+
+    static {
+        String workDir = System.getProperty("user.dir");
```
Comment (trayan-muchev) [inline common/src/main/java/com/market/banica/common/util/ApplicationDirectory.java:17]: You can extract constant for "user.dir"

File: market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java [hunk 3]
```diff
@@ -15,183 +16,191 @@
...
-            this.lock.writeLock().lock();
-
-            String errorMessage = String.format("A good with name %s from %s already exists",
-                    good.toUpperCase(), origin.toUpperCase());
-            String loggerMessage = "Creating and adding a new goodSpecification.";
+            propertiesWriteLock.writeLock().lock();
 
-            origin = origin.toLowerCase(Locale.getDefault());
             good = good.toLowerCase(Locale.getDefault());
+            if (this.doesGoodExist(good)) {
```
Comment (trayan-muchev) [inline market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java:59]: You can add log here.

File: market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java [hunk 4]
```diff
@@ -15,183 +16,191 @@
...
-            this.lock.writeLock().lock();
+            propertiesWriteLock.writeLock().lock();
 
-            origin = origin.toLowerCase(Locale.getDefault());
             good = good.toLowerCase(Locale.getDefault());
 
-            if (!doesGoodExist(String.format(ORIGIN_GOOD_PATTERN, origin, good))) {
-                throw new NotFoundException(String.format("A good with name %s from %s does " +
-                        "not exist and it cannot be removed", good.toUpperCase(), origin.toUpperCase()));
+            if (!doesGoodExist(good)) {
```
Comment (trayan-muchev) [inline market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java:85]: You can add log here.

File: market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java [hunk 5]
```diff
@@ -15,183 +16,191 @@
...
         try {
-            this.lock.writeLock().lock();
+            propertiesWriteLock.writeLock().lock();
 
-            String errorMessage =
-                    String.format("A good with name %s from %s does not exist and cannot be updated", good.toUpperCase(), origin.toUpperCase());
-            String loggerMessage = "Updating an existing goodSpecification.";
-            origin = origin.toLowerCase(Locale.getDefault());
             good = good.toLowerCase(Locale.getDefault());
+            if (!this.doesGoodExist(good)) {
```
Comment (trayan-muchev) [inline market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java:111]: You can add log here.

File: market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java [hunk 6]
```diff
@@ -15,183 +16,191 @@
...
-                , properties, function, this.file, append);
+        this.saveProperties();
 
-        this.goods.put(String.format(ORIGIN_GOOD_PATTERN, origin, good), goodSpecification);
+        this.goods.put(modifiedGoodSpecification.getName(), modifiedGoodSpecification);
 
-        LOGGER.info(loggerMessage);
     }
 
     private void validateGoodSpecification(GoodSpecification goodSpecification) {
```
Comment (trayan-muchev) [inline market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java:141]: Shouldn't we  add a check for negative price here?

File: market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java [hunk 7]
```diff
@@ -15,183 +16,191 @@
...
     }
+
+    private void loadProperties() throws IOException {
+        try (FileInputStream inputStream = new FileInputStream(configurationFile)) {
+            propertiesWriteLock.writeLock().lock();
+
+            properties.load(inputStream);
+
+            List<String> distinctProducts = properties.stringPropertyNames().stream()
+                    .filter(property -> property.matches("([a-z])+\\.([a-z])+\\.([a-z])+"))
```
Comment (trayan-muchev) [inline market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java:182]: You can extract a constant for the regex.

File: market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java [hunk 8]
```diff
@@ -15,183 +16,191 @@
...
+                    .map(property -> property.split("\\.")[0])
+                    .distinct()
+                    .collect(Collectors.toList());
+
+            for (String product : distinctProducts) {
+                GoodSpecification goodSpecification = new GoodSpecification(product,
```
Comment (trayan-muchev) [inline market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java:188]: Extract a method for the GoodSpecification creation. Suggested names (buildGoodSpecification, createGoodSpecification)

File: common/src/main/java/com/market/banica/common/util/ApplicationDirectory.java (hunk 1 above)
Comment (kartaggen) [reply common/src/main/java/com/market/banica/common/util/ApplicationDirectory.java:9]: Renamed it and updated called instances.

File: common/src/main/java/com/market/banica/common/util/ApplicationDirectory.java (hunk 2 above)
Comment (kartaggen) [reply common/src/main/java/com/market/banica/common/util/ApplicationDirectory.java:17]: Extracted the string constant.

File: market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java (hunk 3 above)
Comment (kartaggen) [reply market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java:59]: Added warning log.

File: market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java (hunk 4 above)
Comment (kartaggen) [reply market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java:85]: Added warning log.

File: market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java (hunk 5 above)
Comment (kartaggen) [reply market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java:111]: Added warning log.

File: market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java (hunk 6 above)
Comment (kartaggen) [reply market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java:141]: I must have missed the negative validation, added validation and warning logs to all.

File: market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java (hunk 8 above)
Comment (kartaggen) [reply market-generator/src/main/java/com/market/banica/generator/configuration/MarketConfigurationImpl.java:188]: Extracted method to createGoodSpecification(String goodName).

**Review by Lilyanov — APPROVED** [review]

**Review by trayan-muchev — APPROVED** [review]

**Pull request merged**